### PR TITLE
upwork: init at 5.3.3-883

### PR DESCRIPTION
**Pull request by zakkor**
###### Motivation for this change
I needed to use the Upwork time tracker, and a package for it didn't exist.
Related issue: https://github.com/NixOS/nixpkgs/issues/30654
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/upwork/default.nix [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+, gdk-pixbuf, glib, gtk3, libnotify, libX11, libXcomposite, libXcursor, libXdamage, libuuid
+, libXext, libXfixes, libXi, libXrandr, libXrender, libXtst, nspr, nss, libxcb
+, pango, systemd, libXScrnSaver, libcxx, libpulseaudio }:
+
+stdenv.mkDerivation rec {
+  name = "upwork-${version}";
+  version = "5.3.3-883";
+
+  src = fetchurl {
+    url = https://updates-desktopapp.upwork.com/binaries/v5_3_3_883_1f817bc1fefd44e7/upwork_5.3.3.883_amd64.deb;
```
Comment (LouisTakePILLz) [inline pkgs/applications/misc/upwork/default.nix:12]: Wrap the url in quotes (as per RFC45)

File: pkgs/applications/misc/upwork/default.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  installPhase = ''
+    mv usr $out
+    mv opt $out
+    sed -e "s|/opt/|$out/opt/|g" -i $out/share/applications/upwork.desktop
+
+    wrapProgram $out/opt/Upwork/upwork \
+      "''${gappsWrapperArgs[@]}" \
+      --prefix XDG_DATA_DIRS : "${gtk3}/share/gsettings-schemas/${gtk3.name}/" \
+      --prefix LD_LIBRARY_PATH : ${libPath}
+  '';
```
Comment (jonringer) [inline pkgs/applications/misc/upwork/default.nix:47]: please install the executable in the bin directory as well
```suggestion
      --prefix LD_LIBRARY_PATH : ${libPath}

    mkdir -p $out/bin
    ln -s $out/opt/Upwork/upwork $out/bin/upwork
  '';
```

File: pkgs/applications/misc/upwork/default.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    url = https://updates-desktopapp.upwork.com/binaries/v5_3_3_883_1f817bc1fefd44e7/upwork_5.3.3.883_amd64.deb;
+    sha256 = "072zns79w4h46bvbj23rvr8i12sf2l378ry0z3hchwcimkrph9wx";
+  };
+
+  dontWrapGApps = true;
+
+  nativeBuildInputs = [ 
+    dpkg
+    wrapGAppsHook
+    autoPatchelfHook
```
[the remark this answers is not in the document]
Comment (zakkor) [reply pkgs/applications/misc/upwork/default.nix:21]: As far as I can tell it doesn't work without doing patchelf

**Comment by LouisTakePILLz** [comment]
Running `/opt/Upwork/upwork` in `nix-shell --pure` seems to fail:
```
No protocol specified

(upwork:29262): Gtk-WARNING **: 16:39:07.676: cannot open display: :0
```

Running `/bin/upwork` (the symlink) causes it to segfault in an impure shell:
```
[1]    28927 segmentation fault (core dumped)  /nix/store/y0150qnl83w9ida3ir2al5b5i76nqnyg-upwork-5.3.3-883/bin/upwork
```
In a pure shell however, the symlink gives the same "No protocol specified" error.

**Comment by jonringer** [comment]
> In a pure shell however, the symlink gives the same "No protocol specified" error.
> 
> Tested with `HOME="$(mktemp -d)" nix-shell --pure`

thats because there's not `/etc/protocols` in a sandbox, this is a known-issue see https://github.com/NixOS/nix/issues/1238

it's unfortunate, but just a lot of GUI applications need some impurity to work

**Comment by zakkor** [comment]
I've implemented all your suggestions and used `makeWrapper` directly to create `/bin/upwork` as a workaround for the symlink issue.
(I've also updated the .desktop file to point to the new `/bin/upwork` wrapper so everything works correctly)

**Review by jonringer — APPROVED** [review]
diff LGTM
gui launches fine

```
[2 built, 0.0 MiB DL]
https://github.com/NixOS/nixpkgs/pull/85396
43 package marked as broken and skipped:
clang-sierraHack-stdenv crossLibcStdenv digitalbitbox glibcCross iterm2 linuxPackages-libre.sch_cake linuxPackages.sch_cake linuxPackages_4_14.sch_cake linuxPackages_4_19.sch_cake linuxPackages_4_4.evdi linuxPackages_4_4.sysdig linuxPackages_4_9.sysdig linuxPackages_5_5.sch_cake linuxPackages_5_6.sch_cake linuxPackages_hardened.sch_cake linuxPackages_hardkernel_4_14.bpftrace linuxPackages_hardkernel_4_14.can-isotp linuxPackages_hardkernel_4_14.chipsec linuxPackages_hardkernel_4_14.digimend linuxPackages_hardkernel_4_14.evdi linuxPackages_hardkernel_4_14.mba6x_bl linuxPackages_hardkernel_4_14.nvidia_x11_legacy390 linuxPackages_hardkernel_4_14.nvidiabl linuxPackages_hardkernel_4_14.rtl8814au linuxPackages_hardkernel_4_14.rtl8821ce linuxPackages_hardkernel_4_14.rtlwifi_new linuxPackages_hardkernel_4_14.sch_cake linuxPackages_hardkernel_4_14.v4l2loopback linuxPackages_latest-libre.sch_cake linuxPackages_xen_dom0_hardened.sch_cake muslCross newlibCross octave-jit octoprint php73Extensions.zmq php74Extensions.couchbase php74Extensions.pcs php74Extensions.pthreads python27Packages.caffe python27Packages.libmodulemd python38Packages.nixpart synfigstudio ucommon

1 package built:
upwork
```

**Comment by jonringer** [comment]
@zakkor congrats on your first PR :)

**Pull request merged**